apm-agent-plugins/apm-jakarta-websocket-plugin/src/main/java/co/elastic/apm/agent/websocket/BaseServerEndpointInstrumentation.java

```diff
@@ -0,0 +1,116 @@
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.isInAnyPackage;
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static net.bytebuddy.matcher.ElementMatchers.isAnnotatedWith;
+import static net.bytebuddy.matcher.ElementMatchers.isBootstrapClassLoader;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.namedOneOf;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+
+public abstract class BaseServerEndpointInstrumentation extends TracerAwareInstrumentation {
+
+    private final Collection<String> applicationPackages;
+
+    public BaseServerEndpointInstrumentation(ElasticApmTracer tracer) {
+        applicationPackages = tracer.getConfig(StacktraceConfiguration.class).getApplicationPackages();
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return not(isBootstrapClassLoader())
+            .and(classLoaderCanLoadClass(getServerEndpointClassName()));
+    }
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return isInAnyPackage(applicationPackages, ElementMatchers.<NamedElement>any());
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return isAnnotatedWith(named(getServerEndpointClassName()));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return isAnnotatedWith(
+            namedOneOf("javax.websocket.OnOpen", "jakarta.websocket.OnOpen")
+                .or(namedOneOf("javax.websocket.OnMessage", "jakarta.websocket.OnMessage"))
+                .or(namedOneOf("javax.websocket.OnError", "jakarta.websocket.OnError"))
+                .or(namedOneOf("javax.websocket.OnClose", "jakarta.websocket.OnClose")));
+    }
+
+    protected abstract String getServerEndpointClassName();
+
+    protected static class BaseAdvice {
+
+        protected static Object startTransactionOrSetTransactionName(String signature, String frameworkName, @Nullable String frameworkVersion) {
+            Transaction currentTransaction = tracer.currentTransaction();
+            if (currentTransaction == null) {
+                Transaction rootTransaction = tracer.startRootTransaction(Thread.currentThread().getContextClassLoader());
+                if (rootTransaction != null) {
+                    setTransactionName(rootTransaction, signature, frameworkName, frameworkVersion);
+                    return rootTransaction.activate();
+                }
+            } else {
+                setTransactionName(currentTransaction, signature, frameworkName, frameworkVersion);
+            }
+
+            return null;
+        }
+
+        protected static void endTransaction(Object transactionOrNull, @Advice.Thrown @Nullable Throwable t) {
+            if (transactionOrNull != null) {
```

Comment:
Avoid nesting by returning early
```suggestion
            if (transactionOrNull == null) {
                return;
            }
```